File: 2024/Dec 2024 Python/day 1/day1.py
import heapq
from collections import Counter


def main():
    fileName = "../textfiles/day1.txt"
    part_one(fileName=fileName)
    part_two(fileName=fileName)


def part_one(fileName):
    l1 = []
    l2 = []
    with open(fileName, encoding="utf8") as f:
        lines = f.read().splitlines()
        for line in lines:
            if len(line) > 3:
                a = int(line.split()[0])
                b = int(line.split()[1])
                l1.append(a)
                l2.append(b)

        # i am pretty sure heapify is a min heap
        heapq.heapify(l1)
        heapq.heapify(l2)
        s = 0

        while len(l1) > 0 and len(l2) > 0:
            s += abs(heapq.heappop(l1) - heapq.heappop(l2))

        print("Day One Part One answer:", s)


def part_two(fileName):
    l1 = []
    l2 = []
    with open(fileName, encoding="utf8") as f:
        lines = f.read().splitlines()
        for line in lines:
            if len(line) > 3:
                a = int(line.split()[0])
                b = int(line.split()[1])
                l1.append(a)
                l2.append(b)
        count = Counter(l2)
        ans = 0
        for l in l1:
            if l in count:
                ans += l*count[l]
        print("Day One Part Two answer:", ans)


main()

File: 2019/CPP Feb 2025/Day 1/day1.cpp
#include <iostream>
#include <fstream>
#include <vector>

void partOne(std::string fileName);
void partTwo(std::string fileName);
int getNewValue(int oldValue);
int getNewValuePt2(int oldValue);

int main()
{
    partOne("../textfiles/day1final.txt");
    partTwo("../textfiles/day1final.txt");
    return 0;
}

void partOne(std::string fileName)
{
    // read the lines
    std::ifstream textFile(fileName);
    int num;
    std::vector<int> vals;

    while (textFile >> num)
    {
        // std::cout << num << std::endl;
        vals.push_back(num);
    }

    int s{0};
    for (int v : vals)
    {
        s += getNewValue(v);
    }

    std::cout << "Day 1: part: " << s << std::endl;
}

int getNewValue(int oldValue)
{
    oldValue /= 3;
    oldValue -= 2;
    return oldValue;
}

void partTwo(std::string fileName)
{
    // read the lines
    std::ifstream textFile(fileName);
    int num;
    std::vector<int> vals;

    while (textFile >> num)
    {
        // std::cout << num << std::endl;
        vals.push_back(num);
    }

    int s{0};
    // std::cout << getNewValuePt2(1969) << std::endl;
    for (int v : vals)
    {
        s += getNewValuePt2(v);
    }

    std::cout << "Day 1 part 2: " << s << std::endl;
}

int getNewValuePt2(int oldValue)
{
    if (oldValue < 1)
    {
        return 0;
    }
    int v = getNewValue(oldValue);
    // std::cout << v << std::endl;
    if (v <= 0)
    {
        return 0;
    }
    return v + getNewValuePt2(v);
}